### PR TITLE
[data] log op metrics on completion

### PR DESCRIPTION
**Pull request by Zandew**
## Why are these changes needed?

Logs operator metrics to `ray-data.log` upon operator completion. 
Example:
```
>>> ray.data.range(100).map_batches(lambda x: x).materialize()
2023-10-23 09:20:13,141 INFO streaming_executor.py:305 -- Operator InputDataBuffer[Input] completed. Operator Metrics:
{'num_inputs_received': 20, 'bytes_inputs_received': 45900, 'num_outputs_taken': 20, 'bytes_outputs_taken': 45900, 'cpu_usage': 0, 'gpu_usage': 0}
2023-10-23 09:20:13,592 INFO streaming_executor.py:305 -- Operator InputDataBuffer[Input] -> TaskPoolMapOperator[ReadRange->MapBatches(<lambda>)] completed. Operator Metrics:
{'num_inputs_received': 20, 'bytes_inputs_received': 45900, 'num_inputs_processed': 20, 'bytes_inputs_processed': 45900, 'num_outputs_generated': 20, 'bytes_outputs_generated': 800, 'num_outputs_taken': 20, 'bytes_outputs_taken': 800, 'num_outputs_of_finished_tasks': 20, 'bytes_outputs_of_finished_tasks': 800, 'num_tasks_submitted': 20, 'num_tasks_running': 0, 'num_tasks_have_outputs': 20, 'num_tasks_finished': 20, 'obj_store_mem_alloc': 800, 'obj_store_mem_freed': 45900, 'obj_store_mem_cur': 0, 'obj_store_mem_peak': 22990, 'obj_store_mem_spilled': 0, 'cpu_usage': 0, 'gpu_usage': 0, 'ray_remote_args': {'num_cpus': 1, 'scheduling_strategy': 'SPREAD'}}
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test.py [hunk 1]
```diff
@@ -1,7 +0,0 @@
-import ray
```
Comment (Zandew) [inline python/ray/data/tests/test.py:1]: This was added accidentally in one of my previous prs, removing it here

File: python/ray/data/_internal/execution/streaming_executor.py [hunk 2]
```diff
@@ -83,6 +83,9 @@ def __init__(self, options: ExecutionOptions, dataset_tag: str = "unknown_datase
         self._backpressure_policies: Optional[List[BackpressurePolicy]] = None
 
         self._dataset_tag = dataset_tag
+        # Stores if an operator is completed,
+        # used for marking when an op has just completed.
+        self._op_completed: Optional[Dict[PhysicalOperator, bool]] = None
```
Comment (bveeramani) [inline python/ray/data/_internal/execution/streaming_executor.py:88]: Nit: Might make it clearer that the value type is `bool`.
```suggestion
        self._has_op_completed: Optional[Dict[PhysicalOperator, bool]] = None
```

File: python/ray/data/_internal/execution/streaming_executor.py [hunk 3]
```diff
@@ -293,6 +298,15 @@ def _scheduling_loop_step(self, topology: Topology) -> bool:
             [op.metrics for op in self._topology], {"dataset": self._dataset_tag}
         )
 
+        # Log metrics of newly completed operators.
+        for op in topology:
+            if op.completed() and not self._op_completed[op]:
+                log_str = f"Operator {op} completed. Operator Metrics:\n" + str(
+                    op._metrics.as_dict()
+                )
```
Comment (bveeramani) [inline python/ray/data/_internal/execution/streaming_executor.py:306]: Don't f-strings call `str` implicitly?
```suggestion
                log_str = f"Operator {op} completed. Operator Metrics:\n{op._metrics.as_dict()}"
```

File: python/ray/data/tests/test_stats.py [hunk 4]
```diff
@@ -1234,6 +1234,28 @@ def test_dataset_name():
     )
 
 
+def test_op_metrics_logging():
+    logger = DatasetLogger(
+        "ray.data._internal.execution.streaming_executor"
+    ).get_logger()
+    with patch.object(logger, "info") as mock_logger:
+        ray.data.range(100).map_batches(lambda x: x).materialize()
```
Comment (bveeramani) [inline python/ray/data/tests/test_stats.py:1242]: Nit: If it's not important that we have many rows, `1` might be a better value. Writing minimally passing tests is a good practice. https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#write-minimally-passing-tests
```suggestion
        ray.data.range(1).map_batches(lambda x: x).materialize()
```

**Review by bveeramani — APPROVED** [review]

File: python/ray/data/tests/test_stats.py (hunk 4 above)
Comment (Zandew) [reply python/ray/data/tests/test_stats.py:1242]: Here I want the loop to run more than once just to check if the log is printing exactly once per operator and not every loop. I can probably reduce the size a bit

**Review by raulchen — APPROVED** [review]

**Comment by raulchen** [comment]
retrying the failed test. it doesn't look related.

**Pull request merged**